presto-main/src/test/java/io/prestosql/tests/LogTestDurationListener.java

```diff
@@ -13,30 +13,121 @@
 {
     private static final Logger LOG = Logger.get(LogTestDurationListener.class);
 
     private static final Duration SINGLE_TEST_LOGGING_THRESHOLD = Duration.valueOf("30s");
     private static final Duration CLASS_LOGGING_THRESHOLD = Duration.valueOf("1m");
+    private static final Duration GLOBAL_IDLE_LOGGING_THRESHOLD = Duration.valueOf("5m");
+
+    private final ScheduledExecutorService scheduledExecutorService;
 
     private final Map<String, Long> started = new ConcurrentHashMap<>();
+    private final AtomicLong lastChange = new AtomicLong(System.nanoTime());
+    private final AtomicBoolean hangLogged = new AtomicBoolean();
+    private final AtomicBoolean finished = new AtomicBoolean();
+    @GuardedBy("this")
+    private ScheduledFuture<?> monitorHangTask;
+
+    public LogTestDurationListener()
+    {
+        scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(daemonThreadsNamed("TestHangMonitor"));
+    }
+
+    @Override
+    public synchronized void onExecutionStart()
+    {
+        resetHangMonitor();
+        finished.set(false);
+        if (monitorHangTask == null) {
+            monitorHangTask = scheduledExecutorService.scheduleWithFixedDelay(this::checkForTestHang, 5, 5, TimeUnit.SECONDS);
+        }
+    }
+
+    @Override
+    public synchronized void onExecutionFinish()
+    {
+        resetHangMonitor();
+        finished.set(true);
+        // do not stop hang task so notification of hung threads will fire
+        // Note: since the monitor uses daemon threads it will not prevent JVM shutdown
+    }
+
+    private void checkForTestHang()
+    {
+        if (hangLogged.get()) {
+            return;
+        }
+
+        Duration duration = nanosSince(lastChange.get());
+        if (duration.compareTo(GLOBAL_IDLE_LOGGING_THRESHOLD) < 0) {
+            return;
+        }
+
+        if (!hangLogged.compareAndSet(false, true)) {
+            return;
+        }
+
+        Map<String, Long> runningTests = ImmutableMap.copyOf(started);
+        if (!runningTests.isEmpty()) {
+            String testDetails = runningTests.entrySet().stream()
+                    .map(entry -> String.format("\t%s running for %s", entry.getKey(), nanosSince(entry.getValue())))
+                    .collect(joining("\n"));
+            dumpAllThreads(format("No test started or completed in %s. Running tests:\n%s.", GLOBAL_IDLE_LOGGING_THRESHOLD, testDetails));
+        }
+        else if (finished.get()) {
+            dumpAllThreads(format("Tests finished, but JVM did not shutdown in %s.", GLOBAL_IDLE_LOGGING_THRESHOLD));
+        }
+        else {
+            dumpAllThreads(format("No test started in %s", GLOBAL_IDLE_LOGGING_THRESHOLD));
+        }
+    }
+
+    private static void dumpAllThreads(String message)
+    {
+        LOG.warn("%s\n\nFull Thread Dump:\n%s", message,
+                Arrays.stream(getThreadMXBean().dumpAllThreads(true, true))
```

Comment:
It could be more robust to first log `message` before invoking the thread dump.

Reply:
I'm not sure I follow.  The thread dump here doesn't actually print, it returns objects that we "toString" into the log message, so they all render in one shot.

Reply:
This is good (important) that the thread dump is printed in one shot. I'm just saying `message` could be logged before invoking thread dumping:
```
LOG.warn("%s", message);
LOG.warn("Full Thread Dump:\n%s", ....);
```